substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/pecoff/cv/CVUtil.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (c) 2020, 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, Red Hat Inc. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.  Oracle designates this
+ * particular file as subject to the "Classpath" exception as provided
+ * by Oracle in the LICENSE file that accompanied this code.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package com.oracle.objectfile.pecoff.cv;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+abstract class CVUtil {
```

Comment:
This looks a lot like java.nio.ByteBuffer

Reply:
Yes it does, modulo endianness.

There is a history as to why Simon is not using a ByteBuffer. I followed a similar path with the Dwarf sections implementation classes where the DwarfSectionsImpl.writeXXX methods bottom out at similar looking base DwarfSectionsImpl.putXXX methods.

When trying to size a section the writeXXX methods are passed a null buffer and they compute the size by recursively writing the relevant contents using further writeXXX or putXXX calls and summing the number of bytes written. When a writeXXX method bottoms out at a putXXX method a scratch buffer is passed when the input buffer is null so the computed size reflects the precise number of bytes written. The same top level writeXXX method can then be used to write a buffer allocated with the required total. I thought about using a ByteBuffer when I wrote it this way but decided to keep the byte[] until I saw that there was no need to be able to do raw byte level peeks and pokes on the underlying array.

It would be fairly straightforward to achieve the same outcome by passing a ByteBuffer or null to the writeXXX calls and replacing the putXXX calls with invocations of the relevant ByteBuffer put method. However, what I was planning to do was create a new class, let's call it SectionByteBuffer, with two variants. A StandardSectionByteBuffer would recursively assemble the relevant data into a target byte[] via an encapsulated ByteBuffer. A SizingSectionByteBuffer would throw away the data and simply update a buffer offset to reflect the data written by a putXXX operation.

I was thinking it would be better to fix this for both implementations as a separate patch. Would that be ok? Alternatively, we could make the change to both Linux and Windows debug code now?